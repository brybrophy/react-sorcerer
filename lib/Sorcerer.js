'use strict';

Object.defineProperty(exports, '__esModule', {
	value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _defaultSizes = require('./defaultSizes');

var _defaultSizes2 = _interopRequireDefault(_defaultSizes);

var Sorcerer = (function (_React$Component) {
	_inherits(Sorcerer, _React$Component);

	function Sorcerer(props) {
		_classCallCheck(this, Sorcerer);

		_get(Object.getPrototypeOf(Sorcerer.prototype), 'constructor', this).call(this, props);

		this.addErrorSrc = this.addErrorSrc.bind(this);
	}

	_createClass(Sorcerer, [{
		key: 'addErrorSrc',
		value: function addErrorSrc(e) {
			e.target.src = this.props.errorSrc;
		}
	}, {
		key: 'getSrcSet',
		value: function getSrcSet() {
			var _props = this.props;
			var srcExt = _props.srcExt;
			var srcPath = _props.srcPath;
			var maxDevice = _props.maxDevice;
			var minDevice = _props.minDevice;

			var sizes = this.props.sizes ? this.props.sizes : (0, _defaultSizes2['default'])();
			var min = minDevice ? sizes[minDevice] : 320;
			var max = maxDevice ? sizes[maxDevice] : 3840;
			var srcSetStr = '';

			for (var size in sizes) {
				if (sizes[size] >= min && sizes[size] <= max) {
					if (sizes[size] === max) {
						srcSetStr += srcPath + '_' + size + '.' + srcExt + ' ' + sizes[size] + 'w';
					} else {
						srcSetStr += srcPath + '_' + size + '.' + srcExt + ' ' + sizes[size] + 'w, ';
					}
				}
			}

			return srcSetStr;
		}
	}, {
		key: 'imageProcessor',
		value: (function (_imageProcessor) {
			function imageProcessor() {
				return _imageProcessor.apply(this, arguments);
			}

			imageProcessor.toString = function () {
				return _imageProcessor.toString();
			};

			return imageProcessor;
		})(function () {
			return imageProcessor;
		})
	}, {
		key: 'render',
		value: function render() {
			return _react2['default'].createElement('img', {
				alt: this.props.alt,
				className: this.props.className ? this.props.className : '',
				src: this.props.srcPath + '.' + this.props.srcExt,
				srcSet: this.getSrcSet(),
				onError: this.props.errorSrc ? this.addErrorSrc : null
			});
		}
	}]);

	return Sorcerer;
})(_react2['default'].Component);

exports['default'] = Sorcerer;

Sorcerer.propTypes = {
	alt: _react2['default'].PropTypes.string.isRequired,
	className: _react2['default'].PropTypes.string,
	sizes: _react2['default'].PropTypes.object,
	errorSrc: _react2['default'].PropTypes.string,
	maxDevice: _react2['default'].PropTypes.string,
	minDevice: _react2['default'].PropTypes.string,
	srcExt: _react2['default'].PropTypes.string.isRequired,
	srcPath: _react2['default'].PropTypes.string.isRequired
};
module.exports = exports['default'];