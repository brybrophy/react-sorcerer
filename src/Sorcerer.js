import React, { PropTypes } from 'react';
import { getDevices } from './defaultDevices';

export default class Sorcerer extends React.Component {
	constructor(props) {
		super(props);

		this.addErrorSrc = this.addErrorSrc.bind(this);
	}

	addErrorSrc(e) {
		e.target.src = this.props.errorSrc;
	}

	getSrcSet() {
		const { defaultSrc, maxDevice, minDevice } = this.props;
		const devices = this.props.devices ? this.props.devices : getDevices();
		const srcName = defaultSrc.split('.')[0];
		const srcExt = defaultSrc.split('.')[1];

		let srcSetStr = '';

		const min = minDevice ? devices[minDevice] : 320;
		const max = maxDevice ? devices[maxDevice] : 3840;

		for (let device in devices) {
			if (devices[device] >= min && devices[device] <= max) {
				if (devices[device] === max) {
					srcSetStr += `${srcName}_${device}.${srcExt} ${devices[device]}w`;
				} else {
					srcSetStr += `${srcName}_${device}.${srcExt} ${devices[device]}w, `;
				}
			}
		}

		return srcSetStr;
	}

	render() {
		return (
			<img
				alt={this.props.alt}
				className={this.props.className ? this.props.className : ''}
				src={this.props.defaultSrc}
				srcSet={this.getSrcSet()}
				onError={this.props.errorSrc ? this.addErrorSrc : null}
			/>
		);
	}
}

Sorcerer.propTypes = {
	alt: React.PropTypes.string.isRequired,
	className: React.PropTypes.string,
	defaultSrc: React.PropTypes.string.isRequired,
	devices: React.PropTypes.object,
	errorSrc: React.PropTypes.string,
	maxDevice: React.PropTypes.string,
	minDevice: React.PropTypes.string
}
