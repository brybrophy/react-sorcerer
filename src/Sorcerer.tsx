import React, { CSSProperties } from 'react';

import { useSorcerer, UseSorcererConfig } from './useSorcerer';

export interface ReactSorcererProps extends UseSorcererConfig {
  className?: string;
  style?: CSSProperties;
}

const Sorcerer = (props: ReactSorcererProps) => {
  const sorcererImageProps = useSorcerer({
    alt: props.alt,
    coverage: props.coverage,
    src: props.src,
    srcSetData: props.srcSetData
  });

  return (
    <img
      {...sorcererImageProps}
      className={props.className}
      src={props.src}
      style={props.style}
    />
  );
};

export default Sorcerer;
