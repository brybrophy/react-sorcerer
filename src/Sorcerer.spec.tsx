import React from 'react';
import { shallow } from 'enzyme';

import Sorcerer from '../src/index';
import {
  srcSetData,
  expectedAlt,
  expectedSizes,
  expectedSrc,
  expectedSrcSet
} from '../testUtils/testData';

describe('Sorcerer Component', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(
      <Sorcerer
        alt="an image"
        coverage={100}
        fallbackSrc={expectedSrc}
        srcSetData={srcSetData}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('assigns the correct props', () => {
    const wrapper = shallow(
      <Sorcerer
        alt={expectedAlt}
        coverage={100}
        fallbackSrc={expectedSrc}
        srcSetData={srcSetData}
      />
    );

    expect(wrapper.prop('alt')).toBe(expectedAlt);
    expect(wrapper.prop('sizes')).toBe(expectedSizes);
    expect(wrapper.prop('src')).toBe(expectedSrc);
    expect(wrapper.prop('srcSet')).toBe(expectedSrcSet);
  });
});
