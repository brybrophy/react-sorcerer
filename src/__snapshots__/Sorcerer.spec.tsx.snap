// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorcerer Component matches snapshot 1`] = `ShallowWrapper {}`;
